sdk/servicebus/service-bus/src/core/messageReceiver.ts

```diff
@@ -711,6 +711,57 @@ export class MessageReceiver extends LinkEntity {
     };
   }
 
+  /**
+   * Prevents us from receiving any further messages.
+   */
+  public stopReceivingMessages(): Promise<void> {
+    log.receiver(`[${this.name}] User has requested to stop receiving new messages, attempting to drain the credits.`);
+    this._stopReceivingMessages = true;
+
+    return this.drainReceiver();
+  }
+
+  private drainReceiver(): Promise<void> {
+    log.receiver(`[${this.name}] Receiver is starting drain.`);
+
+    const drainPromise = new Promise<void>((resolve) => {
+      if (this._receiver == null) {
+        log.receiver(`[${this.name}] Internal receiver has been removed. Not draining.`);
+        resolve();
+        return;
+      }
+
+      this._receiver.once(ReceiverEvents.receiverDrained, () => {
+        log.receiver(`[${this.name}] Receiver has been drained.`);
+        resolve();
+      });
+
+      this._receiver.drain = true;
+      // this is not actually adding another credit - it'll just
+      // cause the drain call to start.
+      this._receiver.addCredit(1);
+    });
+
+    return drainPromise;
+  }
+
+  /**
+   * Adds credits to the receiver, respecting any state that
+   * indicates the receiver is closed or should not continue
+   * to receive more messages.
+   *
+   * @param credits Number of credits to add.
+   */
+  protected addCredit(credits: number): boolean {
+    // TODO: seems useful to also migrate a closed check into here.
```

Comment:
The closed check should not be needed I think.
If close() was called on the `MessageReceiver` instance, then `this._receiver` would be undefined.
You could check for `this._receiver.isOpen()`, but if that returned false, then the addition of credits would be a no-op anyway

Reply:
I think you're right about this. When I was adding in this flag I was trying to think about what conditions (that we track manually) 

> 
> 
> What is the current behavior after this change when a user calls `subscribe()` on the receiver after closing the previous `subscribe()` call?
> 
> From what I recall, the helper method `_throwIfAlreadyReceiving()` would throw an error right?

Yes.

Reply:
I've removed the comment for now.